tiflash panic during br restore in getAndMaybeInitStore

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)
v7.5.6
```
[2025/12/10 17:36:45.514 +00:00] [INFO] [BaseDaemon.cpp:1182] ["TiFlash build info: TiFlash\nRelease Version: v7.5.6\nEdition:         Enterprise\nGit Commit Hash: 68d27be0875598b6eec3312778b50521c0c50f12\nGit Branch:      HEAD\nUTC Build Time:  2025-02-27 10:27:58\nEnable Features: jemalloc sm4(GmSSL) asimd unwind thinlto\nProfile:         RELWITHDEBINFO\n"] [thread_id=1]
```

```
[2025/12/10 17:36:27.947 +00:00] [ERROR] [BaseDaemon.cpp:563] ["\n       0x6acd1bc\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+111989180]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:214\n  0xffffb4176850\t<unknown symbol> [linux-vdso.so.1+2128]\n       0x72cd968\tDB::StorageDeltaMerge::getAndMaybeInitStore(DB::ThreadPoolImpl<DB::ThreadFromGlobalPoolImpl<false> >*) [tiflash+120379752]\n                \tdbms/src/Storages/StorageDeltaMerge.cpp:1936\n       0x7c07110\tvoid DB::KVStore::releasePreHandledSnapshot<DB::RegionPtrWithSnapshotFiles>(DB::RegionPtrWithSnapshotFiles const&, DB::TMTContext&) [tiflash+130052368]\n                \tdbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:773\n       0x7be5158\tReleasePreHandledSnapshot [tiflash+129913176]\n                \tdbms/src/Storages/KVStore/FFI/ProxyFFI.cpp:738\n  0xffffb18c5780\t_$LT$engine_store_ffi..observer..TiFlashObserver$LT$T$C$ER$GT$$u20$as$u20$raftstore..coprocessor..ApplySnapshotObserver$GT$::cancel_apply_snapshot::h397b71c26d677a8f [libtiflash_proxy.so+25823104]\n  0xffffb2731e6c\traftstore::store::worker::region::Runner$LT$EK$C$R$C$T$GT$::handle_pending_applies::h05bf95f1570f6c67 [libtiflash_proxy.so+40947308]\n  0xffffb1de2c30\tyatp::task::future::RawTask$LT$F$GT$::poll::h15383bccb4d932ed [libtiflash_proxy.so+31185968]\n  0xffffb34c9214\t_$LT$yatp..task..future..Runner$u20$as$u20$yatp..pool..runner..Runner$GT$::handle::h5adbfadd82cb614e [libtiflash_proxy.so+55198228]\n  0xffffb19bd724\tstd::sys_common::backtrace::__rust_begin_short_backtrace::hcea1723f02df148f [libtiflash_proxy.so+26838820]\n  0xffffb19f4bc4\tcore::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h3aa4e6cbbdc368f4 [libtiflash_proxy.so+27065284]\n  0xffffb2d303c0\tstd::sys::unix::thread::Thread::new::thread_start::hcc916efc5918f503 [libtiflash_proxy.so+47231936]\n  0xffffafce46b8\tstart_thread [libc.so.6+526008]"] [source=BaseDaemon] [thread_id=84]
```

--- a/dbms/src/Common/FailPoint.cpp
+++ b/dbms/src/Common/FailPoint.cpp
@@ -133,7 +133,8 @@ namespace DB
     M(force_join_v2_probe_enable_lm)                         \
     M(force_join_v2_probe_disable_lm)                        \
     M(force_s3_random_access_file_init_fail)                 \
-    M(force_s3_random_access_file_read_fail)
+    M(force_s3_random_access_file_read_fail)                 \
+    M(force_release_snap_meet_null_storage)
 
 #define APPLY_FOR_PAUSEABLE_FAILPOINTS_ONCE(M)    \
     M(pause_with_alter_locks_acquired)            \

--- a/dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp
+++ b/dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp
@@ -37,6 +37,8 @@
 #include <Storages/StorageDeltaMergeHelpers.h>
 #include <TiDB/Schema/SchemaSyncer.h>
 #include <TiDB/Schema/TiDBSchemaManager.h>
+#include <common/logger_useful.h>
+#include <fiu.h>
 
 #include <chrono>
 
@@ -56,6 +58,7 @@ extern const char force_set_parallel_prehandle_threshold[];
 extern const char force_raise_prehandle_exception[];
 extern const char pause_before_prehandle_snapshot[];
 extern const char pause_before_prehandle_subtask[];
+extern const char force_release_snap_meet_null_storage[];
 } // namespace FailPoints
 
 namespace ErrorCodes
@@ -885,10 +888,25 @@ void KVStore::releasePreHandledSnapshot<RegionPtrWithSnapshotFiles>(
     const RegionPtrWithSnapshotFiles & s,
     TMTContext & tmt)
 {
-    auto & storages = tmt.getStorages();
     auto keyspace_id = s.base->getKeyspaceID();
     auto table_id = s.base->getMappedTableID();
+    if (s.external_files.empty())
+    {
+        LOG_INFO(
+            log,
+            "Release prehandled snapshot is skipped, no external files, region_id={} keyspace={} table_id={}",
+            s.base->id(),
+            keyspace_id,
+            table_id);
+        return;
+    }
+
+    auto & storages = tmt.getStorages();
     auto storage = storages.get(keyspace_id, table_id);
+    fiu_do_on(FailPoints::force_release_snap_meet_null_storage, {
+        LOG_WARNING(log, "failpoint force_release_snap_meet_null_storage is enabled");
+        storage = nullptr;
+    });
     if (storage == nullptr)
     {
         LOG_WARNING(

--- a/dbms/src/Storages/KVStore/tests/gtest_raftstore_v2.cpp
+++ b/dbms/src/Storages/KVStore/tests/gtest_raftstore_v2.cpp
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <Storages/KVStore/MultiRaft/ApplySnapshot.h>
 #include <Storages/KVStore/Read/LearnerRead.h>
 #include <Storages/KVStore/tests/region_kvstore_test.h>
 
@@ -22,6 +23,7 @@ namespace FailPoints
 extern const char force_raise_prehandle_exception[];
 extern const char pause_before_prehandle_subtask[];
 extern const char force_set_sst_to_dtfile_block_size[];
+extern const char force_release_snap_meet_null_storage[];
 } // namespace FailPoints
 
 namespace tests
@@ -524,6 +526,120 @@ try
             sp.disable();
             t.join();
         }
+    }
+}
+CATCH
+
+// Test when a region is cancel during releasing pre-handled snapshot.
+// And the pre-handled snapshot contains no external files.
+TEST_F(RegionKVStoreV2Test, KVStoreSingleSnapReleaseNoExternalFiles)
+try
+{
+    auto & ctx = TiFlashTestEnv::getGlobalContext();
+    proxy_instance->cluster_ver = RaftstoreVer::V2;
+    ASSERT_NE(proxy_helper->sst_reader_interfaces.fn_key, nullptr);
+    ASSERT_NE(proxy_helper->fn_get_config_json, nullptr);
+    FailPointHelper::enableFailPoint(FailPoints::force_release_snap_meet_null_storage, static_cast<size_t>(0));
+    SCOPE_EXIT({ FailPointHelper::disableFailPoint(FailPoints::force_release_snap_meet_null_storage); });
+    UInt64 region_id = 2;
+    initStorages();
+    KVStore & kvs = getKVS();
+    TableID table_id = proxy_instance->bootstrapTable(ctx, kvs, ctx.getTMTContext());
+    auto start = RecordKVFormat::genKey(table_id, 0);
+    auto end = RecordKVFormat::genKey(table_id, 40);
+    proxy_instance->bootstrapWithRegion( //
+        kvs,
+        ctx.getTMTContext(),
+        region_id,
+        std::make_pair(start.toString(), end.toString()));
+    auto r1 = proxy_instance->getRegion(region_id);
+
+    auto [value_write, value_default] = proxy_instance->generateTiKVKeyValue(111, 999);
+    auto kkk = RecordKVFormat::decodeWriteCfValue(TiKVValue::copyFrom(value_write));
+    {
+        // mock an empty region
+        MockSSTReader::getMockSSTData().clear();
+        MockSSTGenerator default_cf{region_id, table_id, ColumnFamilyType::Default};
+        default_cf.finish_file(SSTFormatKind::KIND_TABLET);
+        default_cf.freeze();
+        MockSSTGenerator write_cf{region_id, table_id, ColumnFamilyType::Write};
+        write_cf.finish_file(SSTFormatKind::KIND_TABLET);
+        write_cf.freeze();
+        auto [prehandle_region, res] = proxy_instance->snapshot( //
+            kvs,
+            ctx.getTMTContext(),
+            region_id,
+            {default_cf, write_cf},
+            0,
+            0,
+            std::nullopt);
+        kvs.abortPreHandleSnapshot(region_id, ctx.getTMTContext());
+        RegionPtrWithSnapshotFiles region_with_snap(prehandle_region, {});
+        kvs.releasePreHandledSnapshot(region_with_snap, ctx.getTMTContext());
+    }
+}
+CATCH
+
+// Test when a region is cancel during releasing pre-handled snapshot.
+// And the pre-handled snapshot's storage is null.
+TEST_F(RegionKVStoreV2Test, KVStoreSingleSnapReleaseNullStorage)
+try
+{
+    auto & ctx = TiFlashTestEnv::getGlobalContext();
+    proxy_instance->cluster_ver = RaftstoreVer::V2;
+    ASSERT_NE(proxy_helper->sst_reader_interfaces.fn_key, nullptr);
+    ASSERT_NE(proxy_helper->fn_get_config_json, nullptr);
+    FailPointHelper::enableFailPoint(FailPoints::force_release_snap_meet_null_storage, static_cast<size_t>(0));
+    SCOPE_EXIT({ FailPointHelper::disableFailPoint(FailPoints::force_release_snap_meet_null_storage); });
+    UInt64 region_id = 2;
+    initStorages();
+    KVStore & kvs = getKVS();
+    TableID table_id = proxy_instance->bootstrapTable(ctx, kvs, ctx.getTMTContext());
+    auto start = RecordKVFormat::genKey(table_id, 0);
+    auto end = RecordKVFormat::genKey(table_id, 40);
+    HandleID sst_limit = 40;
+    proxy_instance->bootstrapWithRegion( //
+        kvs,
+        ctx.getTMTContext(),
+        region_id,
+        std::make_pair(start.toString(), end.toString()));
+    auto r1 = proxy_instance->getRegion(region_id);
+
+    auto [value_write, value_default] = proxy_instance->generateTiKVKeyValue(111, 999);
+    auto kkk = RecordKVFormat::decodeWriteCfValue(TiKVValue::copyFrom(value_write));
+    {
+        MockSSTReader::getMockSSTData().clear();
+        MockSSTGenerator default_cf{region_id, table_id, ColumnFamilyType::Default};
+        for (HandleID h = 1; h < sst_limit; h++)
+        {
+            auto k = RecordKVFormat::genKey(table_id, h, 111);
+            default_cf.insert_raw(k, value_default);
+        }
+        default_cf.finish_file(SSTFormatKind::KIND_TABLET);
+        default_cf.freeze();
+        MockSSTGenerator write_cf{region_id, table_id, ColumnFamilyType::Write};
+        for (HandleID h = 1; h < sst_limit; h++)
+        {
+            auto k = RecordKVFormat::genKey(table_id, h, 111);
+            write_cf.insert_raw(k, value_write);
+        }
+        write_cf.finish_file(SSTFormatKind::KIND_TABLET);
+        write_cf.freeze();
+
+        auto [prehandle_region, res] = proxy_instance->snapshot( //
+            kvs,
+            ctx.getTMTContext(),
+            region_id,
+            {default_cf, write_cf},
+            0,
+            0,
+            std::nullopt,
+            [] {
+                // set the function so that will call releasePreHandledSnapshot
+                // and verify the null storage handling logic
+                return nullptr;
+            });
+        UNUSED(prehandle_region, res);
     }
 }
 CATCH

